crates/assists/src/handlers/extract_module_to_file.rs

```diff
@@ -0,0 +1,81 @@
+use ast::edit::IndentLevel;
+use ide_db::base_db::{AnchoredPathBuf, FileId};
+use syntax::{
+    ast::{self, edit::AstNodeEdit, NameOwner},
+    AstNode,
+};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: extract_module_to_file
+//
+// This assist extract module to file.
+//
+// ```
+// mod foo {<|>
+//     fn t() {}
+// }
+// ```
+// ->
+// ```
+// mod foo;
+// ```
+pub(crate) fn extract_module_to_file(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let assist_id = AssistId("extract_module_to_file", AssistKind::RefactorExtract);
+    let assist_label = "Extract module to file";
+    let db = ctx.db();
+    let module_ast = ctx.find_node_at_offset::<ast::Module>()?;
+    let module_hir = ctx.sema.to_def(&module_ast)?;
+    let krate = module_hir.krate();
+    let krate_display_name = krate.display_name(db);
+    let krate_name = krate_display_name.as_deref()?;
+    let module_items = module_ast.item_list()?;
+    let dedent_module_items_text = module_items.dedent(IndentLevel(1)).to_string();
+    let module_name = module_ast.name()?;
+    let target = module_ast.syntax().text_range();
+    let new_file_id = FileId(ctx.frange.file_id.0 + 1);
```

Comment:
```suggestion
    let anchor_file_id = FileId(ctx.frange.file_id.0);
```

Anchor isn't the new file. It's a file relative to which the path is treated. Rust-analyzer can't create new files by itself, it can only ask the client (editor) to crate a new file with a specified path. New file will get a file id once the client actually creates it and informs us about that.

Reply:
I see. 

How to implement `create_file` in assist_context module?

In current implementation.

```rust
    pub(crate) fn create_file(&mut self, dst: AnchoredPathBuf, content: impl Into<String>) {
        let file_system_edit = FileSystemEdit::CreateFile { dst: dst.clone() };
        self.file_system_edits.push(file_system_edit);
        self.edit_file(dst.anchor);
        self.insert(TextSize::from(0), content)
    }
```

It creates `CreateFile` and uses edit_file method with `dst.anchor` but it indicates current existed file id not created one. When we want to insert the content to new file, then we get created file id from client. It seems no method to achieve this though.
Any other solution?

Reply:
resolved by https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Fwg-rls-2.2E0/topic/.236746.20Extract.20Module.20to.20File.20assist/near/220129810